### DISCUSSION TITLE
CheckBox color issues

### DISCUSSION THREAD

**Discussion by sedwick2048** [AvaloniaUI/Avalonia#17184] [Q&A] [answered]
I am unable to change the `Foreground` color to something other than Black, and I lose my `Background` color when the box is checked, as though the mouse is still hovering over it.  The issue also happens for tri-state checkboxes--if not in the unchecked state, the background becomes transparent.  A simple user control demonstrates this:

```
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="200" d:DesignHeight="80"
             x:Class="MyApp.Views.CheckBoxIssueTest">
  <StackPanel Background="Red">
    <CheckBox Foreground="White" Background="Green" Content="A checkbox" />
    <CheckBox Foreground="Magenta" Background="Green" Content="A tri-state checkbox" IsThreeState="True" />
  </StackPanel>
</UserControl>
```

![Screenshot 2024-10-03 092712](https://github.com/user-attachments/assets/ebb6dd92-d8f1-4924-b9ec-a2dcbecaa194)
![Screenshot 2024-10-03 092736](https://github.com/user-attachments/assets/82c5cfc6-3a7f-4e50-b3e1-e5d38c4d489b)
![Screenshot 2024-10-03 092752](https://github.com/user-attachments/assets/f86c19b4-1541-4ea3-a2b7-2f44f2846c64)

Much appreciation for any help.  Thanks...

**Comment by timunie** [comment]
You may want to lookup original styles to figure out selectors to use. Often also overriding the DynamicResources can be enough.

**Reply by sedwick2048** [reply]
Where would I look up these "original styles"?  And why doesn't local setting of `Foreground` take effect?

I even added this inside the `StackPanel`:
```
      <StackPanel.Styles>
        <Style Selector="CheckBox">
          <Setter Property="MinHeight" Value="0" />
          <Setter Property="Height" Value="20" />
          <Setter Property="Padding" Value="0" />
          <Setter Property="FontSize" Value="18" />
          <Setter Property="Foreground" Value="White" />
          <Setter Property="Background" Value="Aquamarine" />
          <Setter Property="VerticalContentAlignment" Value="Bottom" />
          <Style Selector="^:pointerover">
            <Setter Property="Background" Value="Yellow" />
          </Style>
        </Style>
      </StackPanel.Styles>
```

`Background` and `Foreground` never take effect, but all the other properties set above do.

**Reply by timunie** [reply, marked as answer]
For example the foreground for unchecked checkbox is defined like this:

https://github.com/AvaloniaUI/Avalonia/blob/master/src%2FAvalonia.Themes.Fluent%2FControls%2FCheckBox.xaml#L75-L77

As you can see its a dynamic resource, so just add this to your App.Resoueces:

```xml
<SolidColorBrush x:Key="KeyToOverride">Yellow</SolidColorBrush>
```

**Reply by sedwick2048** [reply]
Thank you, I created 25 such resource lines, none of which had to be for its background (though I know what to do if I need to):

```
    <!--Unchecked: green box and text; unfilled box-->
    <SolidColorBrush x:Key="CheckBoxForegroundUnchecked">#57AB27</SolidColorBrush>
    <SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeUnchecked">#57AB27</SolidColorBrush>

    <!--Hover over unchecked: box and text turn white-->
    <SolidColorBrush x:Key="CheckBoxForegroundUncheckedPointerOver">White</SolidColorBrush>
    <SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeUncheckedPointerOver">White</SolidColorBrush>

    <!--Pressed unchecked: box and text stay white; box fills green-->
    <SolidColorBrush x:Key="CheckBoxForegroundUncheckedPressed">White</SolidColorBrush>
    <SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeUncheckedPressed">White</SolidColorBrush>
    <SolidColorBrush x:Key="CheckBoxCheckBackgroundFillUncheckedPressed">#57AB27</SolidColorBrush>

    <!--Checked: box and text stay white; box filled green -->
    <SolidColorBrush x:Key="CheckBoxForegroundChecked">White</SolidColorBrush>
    <SolidColorBrush x:Key="CheckBoxCheckBackgroundStrokeChecked">White</SolidColorBrush>
    <SolidColorBrush x:Key="CheckBoxCheckBackgroundFillChecked">#57AB27</SolidColorBrush>
```
...and so on for the checked and indeterminate (middle for tri-state) states, and now it behaves as I expect.  Still wish I could just set `Foreground` and `Background` and have it take effect for all states until specifically overwritten, but hey...

One more little thing, if it's not too much trouble--you notice I have a frequently-used hex value for the green color.  What's the syntax for inserting such a static resource into the above, given its definition and typical uses:

```
    <SolidColorBrush x:Key="SomeGreen" Color="#57AB27" />
```
```
    <TextBlock Text="Some text with a green BG" Background="{StaticResource SomeGreen}" />
```
```
    <Setter Property="Background" Value="{StaticResource SomeGreen}" />
```

**Reply by timunie** [reply]
have a look here: https://github.com/AvaloniaUI/Avalonia/blob/cb4b69894ad8b6e88aa72cb35596813166bbea2c/src/Avalonia.Themes.Fluent/Accents/BaseResources.xaml#L59

**Reply by sedwick2048** [reply]
Sorry for the late reply.  I tried the following in App.axaml:

```
    <SolidColorBrush x:Key="CheckBoxForegroundUnchecked" Color="{StaticResource SomeGreen}" />
```

But it just gave me "Specified cast is not valid", pointing to the axaml file with `CheckBox`es (and not to any line).

**Reply by timunie** [reply]
How is "SomeGreen" defined?

**Reply by sedwick2048** [reply]
In my App.axaml:

```
  <Application.Resources>
    <SolidColorBrush x:Key="MAFGreen" Color="#57AB27" />
  </Application.Resources>
```

**Reply by timunie** [reply]
That's not a color. Corrected: 

```xml
  <Application.Resources>
    <Color x:Key="MAFGreen">#57AB27</Color>
  </Application.Resources>
```

**Reply by sedwick2048** [reply]
Huh, so it wasn't.  It works now, thanks!
